ipld/plugin/nmt.go

```diff
@@ -0,0 +1,406 @@
+	codec uint64,
+	name string,
+	defaultLength int,
+	hashFunc func() hash.Hash,
+) {
+	if _, ok := mh.Codes[codec]; !ok {
+		// make sure that the Codec wasn't registered from somewhere different than this plugin already:
+		if _, found := mh.Codes[codec]; found {
+			panic(fmt.Sprintf("Codec 0x%X is already present: %v", codec, mh.Codes[codec]))
+		}
+		// add to mh.Codes map first, otherwise mh.RegisterHashFunc would err:
+		mh.Codes[codec] = name
+		mh.Names[name] = codec
+		mh.DefaultLengths[codec] = defaultLength
+
+		mhcore.Register(codec, hashFunc)
+	}
+}
+
+type namespaceHasher struct {
+	hash.Hash
+}
+
+func (n namespaceHasher) Sum(data []byte) []byte {
+	isLeafData := data[0] == nmt.LeafPrefix
+	if isLeafData {
+		return nmt.Sha256Namespace8FlaggedLeaf(data[1:])
+	}
+	return nmt.Sha256Namespace8FlaggedInner(data[1:])
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares - 1>
+//
+// Note while this coredag.DagParser is implemented here so this plugin can be used from
+// the commandline, the ipld Nodes will rather be created together with the NMT
+// root instead of re-computing it here.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]ipld.Node, error) {
+	br := bufio.NewReader(r)
+	collector := newNodeCollector()
+
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(collector.visit),
+	)
+
+	for {
+		namespacedLeaf := make([]byte, shareSize+namespaceSize)
+		if _, err := io.ReadFull(br, namespacedLeaf); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if err := n.Push(namespacedLeaf); err != nil {
+			return nil, err
+		}
+	}
+	// to trigger the collection of nodes:
+	_ = n.Root()
+	return collector.ipldNodes(), nil
+}
+
+// nmtNodeCollector creates and collects ipld.Nodes if inserted into a nmt tree.
+// It is mainly used for testing.
+type nmtNodeCollector struct {
+	nodes []ipld.Node
+}
+
+func newNodeCollector() *nmtNodeCollector {
+	// extendedRowOrColumnSize is hardcoded here to avoid importing
+	const extendedRowOrColumnSize = 2 * 128
+	return &nmtNodeCollector{nodes: make([]ipld.Node, 0, extendedRowOrColumnSize)}
+}
+
+func (n nmtNodeCollector) ipldNodes() []ipld.Node {
+	return n.nodes
+}
+
+func (n *nmtNodeCollector) visit(hash []byte, children ...[]byte) {
+	cid := MustCidFromNamespacedSha256(hash)
+	switch len(children) {
+	case 1:
+		n.nodes = prependNode(nmtLeafNode{
+			cid:  cid,
+			Data: children[0],
+		}, n.nodes)
+	case 2:
+		n.nodes = prependNode(nmtNode{
+			cid: cid,
+			l:   children[0],
+			r:   children[1],
+		}, n.nodes)
+	default:
+		panic("expected a binary tree")
+	}
+}
+
+func prependNode(newNode ipld.Node, nodes []ipld.Node) []ipld.Node {
+	nodes = append(nodes, ipld.Node(nil))
+	copy(nodes[1:], nodes)
+	nodes[0] = newNode
```

Comment:
Nit: I feel like this would be an allocation + a copy + a mem _move_. Alternatively, I think simply allocating a new slice of the right capacity then doing a single copy would be more performant, but also easier to read.

Maybe something like:
```go
prepended := make([]ipld.Node, len(nodes)+1)
prepended[0] = newNode
copy(prepended[1:], nodes])
```

Reply:
yes.